src/storage-client/src/controller/remap_migration.rs

```diff
@@ -0,0 +1,343 @@
+        // before processing any updates from the shard.
+        let mut first_ts = T::minimum();
+        first_ts.advance_by(self.since.borrow());
+        native_frontier_updates.push((Partitioned::minimum(), first_ts, 1));
+
+        // Then, we will iterate over the group of updates in time order and produce
+        // the OffsetAntichain at each point, convert it to a normal Antichain, and
+        // diff it with current Antichain representation.
+        for (ts, updates) in &updates.into_iter().group_by(|update| update.1.clone()) {
+            let prev_native_frontier = Antichain::from(compat_source_upper.frontier());
+            native_frontier_updates.extend(
+                prev_native_frontier
+                    .into_iter()
+                    .map(|src_ts| (src_ts, ts.clone(), -1)),
+            );
+
+            compat_source_upper.update_iter(
+                updates
+                    .into_iter()
+                    .map(|(binding, _ts, diff)| (binding, diff)),
+            );
+            let new_native_frontier = Antichain::from(compat_source_upper.frontier());
+
+            native_frontier_updates.extend(
+                new_native_frontier
+                    .into_iter()
+                    .map(|src_ts| (src_ts, ts.clone(), 1)),
+            );
+        }
+
+        // Then, consolidate the native updates and we're done
+        consolidation::consolidate_updates(&mut native_frontier_updates);
+
+        Some((native_frontier_updates, progress))
+    }
+}
+
+#[async_trait::async_trait(?Send)]
+/// Remap handle migration for partitionless sources, e.g. Postgres.
+impl<T: Timestamp + Lattice + Codec64 + TotalOrder> RemapHandleReader
+    for RemapHandleMigrator<MzOffset, T>
+{
+    type FromTime = MzOffset;
+    type IntoTime = T;
+    /// Look through all of the shard's data and return a 0 offset to be written
+    /// if the shard is otherwise empty.
+    async fn next(
+        &mut self,
+    ) -> Option<(
+        Vec<(Self::FromTime, Self::IntoTime, Diff)>,
+        Antichain<Self::IntoTime>,
+    )> {
+        let mut progress = Antichain::from_elem(T::minimum());
+
+        let mut populated = false;
+
+        // Determine whether or not shard is empty and ensure all of its data is
+        // properly formatted as `MzOffset` data.
+        while progress.less_than(self.upper.as_option().unwrap()) {
+            let (bound_updates, progress_inner) = self.subscription.fetch_next().await;
+            for ((update, _), _, _) in bound_updates.into_iter() {
+                populated = true;
+                MzOffset::try_from(update.expect("invalid row")).expect("invalid MzOffset data");
+            }
+            progress = progress_inner;
+        }
+
+        if populated {
+            None
+        } else {
+            Some((
+                vec![(
+                    MzOffset { offset: 0 },
+                    self.since.as_option().unwrap().clone(),
```

Comment:
I'm only leaving this comment for educational purposes. What you want to express here is that you want the shard to be at the same state as if it contained `(MzOffset::minimum(), T::minimum(), 1)` and and then got compacted to `since`.

So what you want to figure out here is "what is the minimum timestamp under a given `since` frontier?", which you can answer by using `Lattice::advance_by` like so:

```rust
let mut min_ts = T::minimum();
min_ts.advance_by(since.borrow());
Some((vec![(MzOffset::minimum(), min_ts, 1)], progress))
```

Reply:
Thanks for this; hadn't thought of using the `Lattice::advance_by`.